File: app/src/main/java/com/agotakiss/androidtest/presentation/detail/SimilarMovieAdapter.java
package com.agotakiss.androidtest.presentation.detail;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import com.agotakiss.androidtest.R;
import com.agotakiss.androidtest.domain.models.Movie;
import com.agotakiss.androidtest.presentation.home.OnEndReachedListener;
import com.squareup.picasso.Picasso;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.agotakiss.androidtest.presentation.home.MovieAdapter.IMAGE_BASE_URL;
import static com.agotakiss.androidtest.presentation.home.MovieAdapter.LAST_ITEMS_BEFORE_LOAD_NEW;
import static com.agotakiss.androidtest.presentation.home.MovieAdapter.MOVIE;

public class SimilarMovieAdapter extends RecyclerView.Adapter<SimilarMovieAdapter.ViewHolder> {
    private List<Movie> similarMovies;
    private Context context;
    private OnEndReachedListener onEndReachedListener;

    public void setOnEndReachedListener(OnEndReachedListener onEndReachedListener) {
        this.onEndReachedListener = onEndReachedListener;
    }

    public SimilarMovieAdapter(List<Movie> similarMovies, Context context) {
        this.similarMovies = similarMovies;
        this.context = context;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.similar_movie_poster) ImageView similarMoviePoster;
        @BindView(R.id.similar_movie_rating) TextView similarMovieRating;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

//            itemView.setOnClickListener(v -> {
//                int itemPosition = getAdapterPosition();
//                Movie similarMovie = similarMovies.get(itemPosition);
//                Intent intent = new Intent(context, DetailsActivity.class);
//                intent.putExtra(MOVIE, similarMovie);
//                ((Activity) context).finish();
//                context.startActivity(intent);
//            });
        }
        @OnClick(R.id.similar_movie_cardview)void onItemClicked(){
            int itemPosition = getAdapterPosition();
            Movie similarMovie = similarMovies.get(itemPosition);
            Intent intent = new Intent(context, DetailsActivity.class);
            intent.putExtra(MOVIE, similarMovie);
            ((Activity) context).finish();
            context.startActivity(intent);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.similar_movie_list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Movie movie = similarMovies.get(position);
        if (position == similarMovies.size() - LAST_ITEMS_BEFORE_LOAD_NEW){
            onEndReachedListener.onEndReached(position);
        }
        holder.similarMovieRating.setText(Float.toString(movie.getAverageVote()));
        Picasso.get().load(IMAGE_BASE_URL + movie.getPosterPath()).into(holder.similarMoviePoster);
    }

    @Override
    public int getItemCount() {
        return similarMovies.size();
    }
}

File: app/src/main/java/com/agotakiss/androidtest/presentation/detail/SimilarMovieAdapter.kt
package com.agotakiss.androidtest.presentation.detail

import android.support.v4.view.ViewCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.agotakiss.androidtest.R
import com.agotakiss.androidtest.domain.models.Movie
import com.agotakiss.androidtest.presentation.POSTER_TRANSITION_NAME
import com.agotakiss.androidtest.presentation.main.MovieAdapter.Companion.IMAGE_BASE_URL
import com.agotakiss.androidtest.presentation.main.MovieAdapter.Companion.LAST_ITEMS_BEFORE_LOAD_NEW
import com.agotakiss.androidtest.presentation.main.OnEndReachedListener
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.similar_movie_list_item.view.*

class SimilarMovieAdapter(private val similarMovies: List<Movie>,
                          private var onEndReachedListener: OnEndReachedListener,
                          private val onItemClickListener: (Movie, View) -> Unit)

    : RecyclerView.Adapter<SimilarMovieAdapter.ViewHolder>() {

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            itemView.setOnClickListener(object : View.OnClickListener {
                override fun onClick(v: View?) {
                    ViewCompat.setTransitionName(itemView.similar_movie_poster, POSTER_TRANSITION_NAME)

                    onItemClickListener.invoke(similarMovies[adapterPosition], itemView.similar_movie_poster)
                }
            })
        }

        fun bindViewHolder(position: Int) {
            val similarMovie = similarMovies[position]
            if (position == similarMovies.size - LAST_ITEMS_BEFORE_LOAD_NEW) {
                onEndReachedListener.onEndReached(position)
            }
            if (similarMovie.averageVote != 0F){
                itemView.similar_movie_rating!!.text = java.lang.Float.toString(similarMovie.averageVote)
            } else{
                itemView.similar_movie_rating.text = "Unrated"
            }
            if (similarMovie.posterPath != null){
                Glide.with(itemView.context)
                    .load(IMAGE_BASE_URL + similarMovie.posterPath)
                    .into(itemView.similar_movie_poster)
            }else{
                itemView.similar_movie_poster.setImageResource(R.drawable.default_poster)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.similar_movie_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindViewHolder(position)
    }

    override fun getItemCount(): Int {
        return similarMovies.size
    }
}
